source/docs/software/wpilib-tools/robot-simulation/device-sim.rst

```diff
@@ -0,0 +1,117 @@
+Device Simulation
+=================
+
+WPILib provides a way to manage simulation device data in the form of the SimDevice API.
+
+Simulating Core WPILib Device Classes
+-------------------------------------
+
+Core WPILib device classes (i.e ``Encoder``, ``Ultrasonic``, etc.) have simulation classes named ``EncoderSim``, ``UltrasonicSim``, etc. These classes allow interactions with the device data that wouldn't be possible or valid outside of simulation.
+
+.. note:: This example will use the ``EncoderSim`` class as an example. Use of other simulation classes will be almost identical.
+
+.. important:: Simulation classes will do nothing on a real robot!
+
+Creating Simulation Device objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+All simulation device classes have a constructor that accepts the regular object, and an additional constructor or factory method to create a simulation object by the ports the device is connected to. The latter is especially useful for :doc:`unit testing <unit-testing>`.
+
+.. tabs::
+   .. code-tab:: java
+
+      // create a real encoder object on DIO 2,3
+      Encoder encoder = new Encoder(2, 3);
+      // create a sim controller for the encoder
+      EncoderSim simEncoder = new EncoderSim(encoder);
+
+   .. code-tab:: cpp
+
+       // create a real encoder object on DIO 2,3
+       Encoder encoder{2, 3};
+       // create a sim controller for the encoder
+       EncoderSim simEncoder{encoder};
+
+Reading and Writing Device Data
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Each simulation class has getter (``getXxx()``/``GetXxx()``) and setter (``setXxx(value)``/``SetXxx(value)``) functions for each field ``Xxx``. The getter functions will return the same as the getter of the regular device class.
+
+.. tabs::
+   .. code-tab:: java
+
+      simEncoder.setCount(100);
+      encoder.getCount(); // 100
+      simEncoder.getCount(); // 100
+
+   .. code-tab:: cpp
+
+      simEncoder.SetCount(100);
+      encoder.GetCount(); // 100
+      simEncoder.GetCount(); // 100
+
+Registering Callbacks
+^^^^^^^^^^^^^^^^^^^^^
+
+In addition to the getters and setters, each field also has a ``registerXxxCallback()`` function that registers a callback to be run whenever the field value changes. These functions return a ``CallbackStore`` object, the callback can be canceled by calling ``close()`` in Java or by automatic destruction in C++. The callbacks accept a string parameter of the name of the field and a ``HALValue`` object containing the value. The value can be retrieved with ``getInt()`` etc.
+
+.. warning:: In C++, do **not** invoke the destructor explicitly! It is invoked automatically on scope exit and causes undefined behavior if explicitly destroyed earlier.
```

Comment:
Having 3 warnings in a row really makes them underpreciate (less impactful). Either put them all in one warning, space them out, or make them not warnings.